src/main/java/io/xstefank/wildfly/bot/util/PullRequestMergableProcessor.java

```diff
@@ -0,0 +1,172 @@
+                    LOGGER.infof("Applying '%s' for Pull Request #%d", LABEL_NEEDS_REBASE, pullRequest.getNumber());
+                    if (!pullRequest.getMergeable()) {
+                        labelsToAdd.add(LABEL_NEEDS_REBASE);
+                    } else {
+                        labelsToRemove.add(LABEL_NEEDS_REBASE);
+                    }
+                }
+
+                return Uni.createFrom().voidItem().invoke(() -> {
+                    try {
+                        githubProcessor.updateLabels(pullRequest, labelsToAdd, labelsToRemove);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        };
+
+        Uni<List<GHPullRequest>> uni = Uni.createFrom()
+                // Collect all Pull Requests
+                .item(repository.queryPullRequests().state(GHIssueState.OPEN).base("main").list())
+                .invoke(() -> LOGGER.infof("Scheduling a merge update for open pull requests for new head [%s - \"%s\"]",
+                        headCommit.getSha(), headCommit.getMessage()))
+                .map(ghPullRequests -> {
+                    try {
+                        return ghPullRequests.toList();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                // Prompt GitHub to recalculate mergable
+                .call(combineUnis(pollGitHub::apply)::apply)
+                // Give GitHub some time
+                .onItem().delayIt().by(Duration.ofSeconds(wildFlyBotConfig.timeout()))
+                // Retrieve the results from GitHub and apply labels accordingly
+                .call(combineUnis(pollGitHub::apply)::apply)
+                // Filter failed Pull Requests
+                .call(combineUnis(applyLabels::apply)::apply);
+
+        pushPayloadsQueue.add(uni);
+        subscription(null, null);
+    }
+
+    /**
+     * Subscribes to the Uni<List<GHPullRequest>> and after the execution it subscribes to the next
+     * such Uni, if available
+     *
+     * @param ghPullRequests List<GHPullRequest> return from the Uni
+     * @param headCommit Commit from the pushPayload prompting this invocation
+     */
+    private void subscription(List<GHPullRequest> ghPullRequests, GHEventPayload.Push.PushCommit headCommit) {
+        if (ghPullRequests != null && headCommit != null) {
+            logResult(ghPullRequests, headCommit);
+        }
+        synchronized (this) {
+            if (pushPayloadsQueue.peek() != null && !currentlyExecuting) {
+                currentlyExecuting = true;
+                pushPayloadsQueue.poll().subscribe().with(list -> {
+                    currentlyExecuting = false;
+                    this.subscription(list, headCommit);
+                });
```

Comment:
I believe the `subscribe().with()` gets executed on the event loop, so it should not be blocking the `synchronize` block then yes. So it will not allow if new request came in to execute it's `Uni` and recursively it gets called by previous finished `Uni` in the `with` method. The recursion ends on empty queue (`Queue#poll` pops the element from the queue).

Reply:
Also to add, I was thinking about adding a `synchronized` inside the `with` method because of the boolean, so no race condition could happen. Tests are passing with this and it ensures even switching from `@Singleton` would not result in possible race conditions